### PR TITLE
LNX-83-create-generic-requirements

### PR DESCRIPTION
**Pull request by Krystian030**
- Common requirements package has been created 
- The `requirements` method has been added to the action interface, which returns a list of requirements that need to be met
- Unit tests have been added
- The first requirement for the move action has been added. More requirements will add after creating generic actions.

How does it work?
Example for Move action. Here, we define the requirements function as follows:
```
  def requirements(self) -> List[Callable[['Scene'], bool]]:
        return [CommonRequirements.is_walkable_destination(self.object_id, self.vector)]
``` 
Inside `CommonRequirements`, we can define requirements for all actions, like so:
```
   @staticmethod
    def is_walkable_destination(object_id: int, vector: Vector):
        def requirement(scene: 'Scene') -> Callable[['Scene'], bool]:
            object: Object = scene.get_object_by_id(object_id)
            destination_position: Vector = object.position + vector
            object_at_destination: Object = scene.get_object_by_position(destination_position)
            return object_at_destination is not None and object_at_destination.walkable

        return requirement
```
If we want the requirement to apply only to our action, we can define it in our `Action` class as follows:
```
  def requirements(self) -> List[Callable[['Scene'], bool]]:
        def requirement(scene: 'Scene') -> Callable[['Scene'], bool]:
            object: Object = scene.get_object_by_id(self.object_id)
            destination_position: Vector = object.position + self.vector
            object_at_destination: Object = scene.get_object_by_position(destination_position)
            return object_at_destination is not None and object_at_destination.walkable
 
        return [requirement]
```

### CODE REVIEW

File: lynx/common/actions/action.py [hunk 1]
```diff
@@ -12,3 +15,8 @@ def apply(self, scene: 'Scene') -> None:
         """
         Method used to change state of a `scene`
         """
+
+    def requirements(self) -> List[Callable[['Scene'], bool]]:
+        """
+        Method return all requirements needed to apply changes
```
Comment (filip-szweda) [inline lynx/common/actions/action.py]: > returns

File: lynx/common/actions/common_requirements.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from typing import Callable
+
+from lynx.common.object import Object
+from lynx.common.vector import Vector
+
+
+class CommonRequirements:
+
+    @staticmethod
+    def is_walkable_destination(object_id: int, vector: Vector):
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]: What is this `vector` exactly?
I think we should rename it. In `Move`, also.

File: lynx/common/scene.py [hunk 3]
```diff
@@ -25,13 +23,13 @@ def add_entity(self, entity: Entity) -> NoReturn:
             self._object_position_map[entity.position] = entity
             self._object_id_map[entity.id] = entity
 
-    def get_object_by_id(self, id: int) -> Union[Object, None]:
+    def get_object_by_id(self, id: int) -> Optional[Object]:
```
Comment (filip-szweda) [inline lynx/common/scene.py:26]: Great call on that xD

File: setup.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/bin/python
 
-from setuptools import setup
+from setuptools import find_packages, setup
```
Comment (filip-szweda) [inline setup.py]: Revert this change.

File: lynx/common/actions/common_requirements.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from typing import Callable
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]: Consider using something like:
```python
class CommonRequirements:
    @staticmethod
    def is_walkable_destination(scene: 'Scene', object_id: int, vector: Vector) -> bool:
        object: Object = scene.get_object_by_id(object_id)
        destination_position: Vector = object.position + vector
        object_at_destination: Object = scene.get_object_by_position(destination_position)
        return object_at_destination is not None and object_at_destination.walkable
...

def satisfies_requirements(self) -> bool:
        return CommonRequirements.is_walkable_destination(self.object_id, self.vector) and ... and ...
```

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: lynx/common/actions/common_requirements.py (hunk 2 above)
Comment (alpinus4) [inline lynx/common/actions/common_requirements.py]: could we change it to more generic requirement
```python
is_walkable(position: Vector)
```
so that we can use it with any action that has to do with walkability

**Review by alpinus4 — CHANGES_REQUESTED** [review]

File: tests/move_test.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import unittest
```
Comment (blazej-smorawski) [inline tests/move_test.py]: We used pytest I think, it's not the same thing is it?

File: lynx/common/actions/common_requirements.py (hunk 5 above)
Comment (blazej-smorawski) [reply lynx/common/actions/common_requirements.py]: Well, it's the same but I like the closure logic much more, for me it's much more elegant. I'd leave it up to Krystian's taste.

**Review by blazej-smorawski — APPROVED** [review]

File: tests/move_test.py (hunk 6 above)
Comment (Krystian030) [reply tests/move_test.py]: Good point!

**Review by EasyJB — APPROVED** [review]

File: lynx/common/actions/action.py (hunk 1 above)
Comment (Krystian030) [reply lynx/common/actions/action.py]: Sorry I don't understand, could you explain?

File: setup.py (hunk 4 above)
Comment (Krystian030) [reply setup.py]: done

File: lynx/common/actions/action.py (hunk 1 above)
Comment (blazej-smorawski) [reply lynx/common/actions/action.py]: "Method return all requirements" -> "Method returns all requirements" I guess

File: lynx/common/actions/common_requirements.py (hunk 2 above)
Comment (blazej-smorawski) [reply lynx/common/actions/common_requirements.py]: Why would you rename vector inside Move? Moving something by some vector is something that you'll see everywhere.

Comment (Krystian030) [reply lynx/common/actions/common_requirements.py]: Rename to `movement`

**Review by filip-szweda — APPROVED** [review]
